src/main/java/com/iota/iri/utils/thread/ThreadUtils.java

```diff
@@ -0,0 +1,139 @@
+     */
+    public static Thread spawnThread(Runnable runnable, ThreadIdentifier threadIdentifier) {
+        if (threads.get(threadIdentifier) == null || threads.get(threadIdentifier).isInterrupted()) {
+            synchronized(threadIdentifier) {
+                if (threads.get(threadIdentifier) == null || threads.get(threadIdentifier).isInterrupted()) {
+                    threads.put(threadIdentifier, spawnThread(runnable, threadIdentifier.getName()));
+                }
+            }
+        }
+
+        return threads.get(threadIdentifier);
+    }
+
+    /**
+     * This method spawns a new {@link Thread} with the given name.
+     *
+     * This method does not actively manage the {@link Thread}s and calling it multiple times will create multiple
+     * {@link Thread}s. It is internally used by {@link #spawnThread(Runnable, ThreadIdentifier)}.
+     *
+     * In addition to creating the {@link Thread}, it also dumps some log messages to inform the user about the creation
+     * and the lifecycle of the {@link Thread}s.
+     *
+     * @param runnable method reference that is responsible for processing the thread (i.e. {@code this::myThreadMethod})
+     * @param threadName the name of the {@link Thread} that shall be started started
+     * @return the thread that got spawned by this method
+     */
+    public static Thread spawnThread(Runnable runnable, String threadName) {
+        logger.info("Starting Thread: " + threadName + " ...");
+
+        Thread thread = new Thread(() -> {
+            logger.info(threadName + " [STARTED]");
+
+            runnable.run();
+
+            logger.info(threadName + " [STOPPED]");
+        }, threadName);
+        thread.start();
+
+        return thread;
+    }
+
+    /**
+     * This method stops the {@link Thread} that is associated with the given {@link ThreadIdentifier}.
+     *
+     * This method is thread safe and only calls the {@link Thread#interrupt()} method once per running {@link Thread}.
+     * If no running {@link Thread} is associated with the given {@link ThreadIdentifier}, it will have no affect.
+     *
+     * It can therefore be used as a convenient way to offer a {@code shutdown()} method in the corresponding manager
+     * instance, without having to worry about starting multiple {@link Thread}s when calling the {@code shutdown()}
+     * method multiple times.
+     *
+     * @param threadIdentifier the identifier of the thread that we want to stop
+     * @return the {@link Thread} that will be terminated or null if there was no {@link Thread} associated to the given
+     *         identifier
+     */
+    public static Thread stopThread(ThreadIdentifier threadIdentifier) {
+        if (threads.get(threadIdentifier) != null && !threads.get(threadIdentifier).isInterrupted()) {
+            synchronized(threadIdentifier) {
+                if (threads.get(threadIdentifier) != null && !threads.get(threadIdentifier).isInterrupted()) {
```

Comment:
Same here.

Reply:
I don't think that using a ConcurrentHashMap will be sufficient because you are also writing to the map after which also have to be synchronized. So you want to synchronize those 2 calls rather than just have the get synchronized.

And in addition the stopping of the thread also have to be synchronized over the same object. Otherwise the following could happen.

You call spawnThread and immediately after stopThread:

1. spawnThread determines there is no running job
2. stopThread determines there is no running job and terminates
3. spawnThread creates a job

With the current implementation synchronizing both calls over the same object:

1. spawnThread determines there is no running job
2. spawnThread creates a job
3. stopThread terminates it